SuperchainERC20 FMA: update doc (#187)
ethereum-optimism/design-docs: security/fma-superchainerc20.md

- [**SuperchainERC20 standard-only FMA (Failure Modes and Recovery Path Analysis)**](#superchainerc20-standard-only-fma-failure-modes-and-recovery-path-analysis)
- [Introduction](#introduction)
  - [Security Considerations](#security-considerations)
- [Failure Modes and Recovery Paths](#failure-modes-and-recovery-paths)
  - [FM1: Unauthorized Access to `crosschainMint` & `crosschainBurn` Functions](#fm1-unauthorized-access-to-crosschainmint--crosschainburn-functions)
  - [FM2: Token contract not Deployed in Destination Chain, but Able to Deploy](#fm2-token-contract-not-deployed-in-destination-chain-but-able-to-deploy)
  - [FM3: Token Deployer is Lost, or Unable to Deploy to Expected Address](#fm3-token-deployer-is-lost-or-unable-to-deploy-to-expected-address)
  - [FM4: Compromised Deployment Method](#fm4-compromised-deployment-method)
- [Action Items](#action-items)
- [Audit Requirements](#audit-requirements)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## **SuperchainERC20 standard-only FMA (Failure Modes and Recovery Path Analysis)**

| Author | Ng, Joxes, Particle, Gotzen |
| --- | --- |
| Created at | 2024-10-02 |

> | Needs Approval From | Mark Tyneway, Matt Solomon |

Reviewer: Would be good to also get a quick review from someone on interop who is familiar with this for any big issues, @tynes could be good but could also be someone from wonderland

Reviewer: Wonderland has reviewed internally, so resolving

- **Description:** The `crosschainMint` and `crosschainBurn` functions can only be called by the `SuperchainTokenBridge`,  enforced by the check `msg.sender != Predeploys.SUPERCHAIN_TOKEN_BRIDGE`. If the bridge address is badly defined or the modifier bypassed, an entity could mint and burn tokens.
- **Risk Assessment**: Medium.
    - Potential impact: High. All tokens based on this implementation could be potentially at risk.
    - Likelihood: Very Low. `Predeploys.SUPERCHAIN_TOKEN_BRIDGE` is defined via protocol upgrades. The conditional is sufficiently simple and battle-tested to give confidence in the implementation.
- **Mitigation**: Current unit tests perform authorization checks for `crosschainMint` ([test](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/test/L2/SuperchainERC20.t.sol#L38)) and `crosschainBurn` ([test](https://github.com/ethereum-optimism/optimism/blob/1add88a41b8cdc5c488a18976e3e4fa4b02e7da4/packages/contracts-bedrock/test/L2/SuperchainERC20.t.sol#L77)). There should be tests to ensure that the `SuperchainTokenBridge` predeploy is correctly set during the interop fork upgrade, and is not subject to unexpected changes.

> - **Detection**: Ideally, run an off-chain agent to monitor every `crosschainMint` and `crosschainBurn` event in real-time, perhaps it could be resource-intensive given the permissionless usage of `SuperchainERC20` and the expected large number of deployments. If that’s not feasible, rely on user-filed support tickets to flag unusual mint/burn activity, which the security team can investigate for suspected unauthorized access.

Reviewer: ```suggestion
- **Detection**: Ideally, run an off-chain agent to monitor every `crosschainMint` and `crosschainBurn` event in real-time, perhaps it could be resource-intensive given the permissionless usage of `SuperchainERC20` and the expected large number of deployments. If that’s not feasible, rely on user-filed support tickets to flag unusual mint/burn activity, which the TBD team (see corresponding action item) can investigate for suspected unauthorized access.
```

- **Recovery Path(s)**: Equivocation (i.e. another implementation being set) on the `SuperchainTokenBridge` address would require a protocol upgrade or hard fork to restore the expected code.

> ### FM2: Token contract not Deployed in Destination Chain, but Able to Deploy

Reviewer: Clarify phrasing to say this failure mode is that the relay fails (failure mode) until the token is deployed (mitigation) at which time the relay can be executed again to solve the problem

- **Description**: When a cross-chain transfer is initiated (via `sendERC20`), the destination chain must have the  `SuperchainERC20` contract deployed at the same address. If the contract does not exist on the destination chain, the `relayERC20`  cannot mint the tokens and thus will fail.
- **Risk Assessment**: Medium

>     - Potential impact: Medium. Users can lose access to their tokens temporarily or permanently if the deployer (in the case of permissioned tokens) doesn’t deploy the token afterwards.

Reviewer: @0xParticle mentioned that the permissioned case will be moved to a different FMA

    - Likelihood: Medium. A deployer may choose not to deploy the token on all chains.
- **Mitigation**: Trusted bridge interfaces should prevent users from sending tokens to a chain where the token doesn’t exist. Double check other trusted sources (such as Superchain Token List) for greater confidence.
- **Detection**: Support tickets filed by users reporting the issue.
- **Recovery Path(s)**: Deploy the token.

> ### FM3: Token Deployer is Lost, or Unable to Deploy to Expected Address

Reviewer: Clarify this is a variant of FM2 where you can't deploy the token on the other chain therefore funds are lost

> ### FM3: Token Deployer is Lost, or Unable to Deploy to Expected Address

Author: I added this in FM2

- **Risk Assessment**: Medium
    - Potential impact: High. In the worst-case scenario, compromised methods open the door to malicious behavior, e.g. where an attacker can occupy the address, and forge bridging events, and gain the ability to mint tokens in other chains.
    - Likelihood: Low. This is possible if the deployer engages in poor practices or lacks opsec. Two examples of this are:
        - Custom factories that utilize CREATE and share the same address are used, where tokens are deployed using the same bytecode and nonce.
        - Proxy contracts where the admin is compromised or the proxy address front-runned.
- **Mitigation**: For developers, employ the appropriate deterministic deployment tools, such as the one at `create2Deployer`. In the case of permissioned deployments, common for proxy contracts, ensure to maintain control over those deployments and ownership.
- **Detection**: Support tickets filed by users reporting such issues.

> - **Recovery Path(s)**: Pause the token contract where possible and proceed to redeploy.

Reviewer: same phrasing as above
```suggestion
- **Recovery Path(s)**: Pause the token contract where possible and proceed to redeploy and migrate userbase if it is needed.
```

## Action Items

The following action items need to be done:

> - [ ]  Resolve all the comments.

Reviewer: ```suggestion
- [ ]  Resolve all the comments.
- [ ] Communicate to bridge frontends to ensure they prevent users from sending tokens to a chain where the token doesn’t exist
```

Author: Included in action items https://github.com/ethereum-optimism/design-docs/blob/df5cd73c5a86bc6abced92f9ef72778c89cf22a1/security/fma-superchainerc20.md?plain=1#L114

> - [ ]  Implement tests for `SuperchainTokenBridge` predeploy. Currently, the L2 Genesis project is under development and it should take care of these tests.

Reviewer: ```suggestion
- [ ]  Implement tests for `SuperchainTokenBridge` predeploy. Currently, the L2 Genesis project is under development and it should take care of these tests (FM1)
```

> - [ ]  Decide whether to use off-chain scripts or rely on a user-support system for FM1.

Reviewer: ```suggestion
- [ ]  Decide whether to use off-chain scripts or rely on a user-support system for FM1.
- [ ] Ensure docs for SuperchainERC20 developers explain the need for deterministic deployment and how to achieve it (FM3)
```

Author: Included in action items https://github.com/ethereum-optimism/design-docs/blob/df5cd73c5a86bc6abced92f9ef72778c89cf22a1/security/fma-superchainerc20.md?plain=1#L115

> - [ ]  Ensure the support team is aware of these failure modes and prepared to respond.

Reviewer: We should expand this to mention that we need to figure out who the support team (i.e. the team getting and responding to alerts) is. Ideally this is the team that wrote and owns the feature, but is that feasible if wonderland isn't part of labs for any infra reasons? Or do we need to find an owner within labs for these alerts?

Reviewer: ```suggestion
- [ ]  Ensure the support team is aware of these failure modes and prepared to respond. Ideally this is the team that wrote and owns the feature, but is that feasible if wonderland isn't part of labs for any infra reasons? Or do we need to find an owner within labs for these alerts?
```

Reviewer 2: @mds1 - I've spoken to Wonderland and we would need some in OP Labs to own these alerts.  I will DM you as well but the pressing concern is to try to get any code changes necessary for this in these contracts before Feb 10th.

Author: Included in action items. Leaving this conversation open for now just for visibility.

## Audit Requirements

> An audit of interop contracts, including the `SuperchainERC20` contract, is expected. However, developers may use `SuperchainERC20` before the interop fork is live.

Reviewer: Let's move this up to the intro

  - [FM2: Token Contract Missing in Destination Chain — Relay Fails Until Deployed (But Deployment Is Permissionless)](#fm2-token-contract-missing-in-destination-chain--relay-fails-until-deployed-but-deployment-is-permissionless)
  - [FM3: Token Contract Missing in Destination Chain — Relay Fails Until Deployed (But Deployment Is Permissioned)](#fm3-token-contract-missing-in-destination-chain--relay-fails-until-deployed-but-deployment-is-permissioned)
  - [FM4: Token Contract Missing in Destination Chain — Token Deployer Is Lost or Unable to Deploy to Expected Address](#fm4-token-contract-missing-in-destination-chain--token-deployer-is-lost-or-unable-to-deploy-to-expected-address)
  - [FM5: Compromised Deployment Method](#fm5-compromised-deployment-method)
- [Action Items](#action-items)
<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## **SuperchainERC20 standard-only FMA (Failure Modes and Recovery Path Analysis)**

| Author | Ng, Joxes, Particle, Gotzen |
| --- | --- |
| Created at | 2024-10-02 |
| Needs Approval From | Mark Tyneway, Matt Solomon |
| Other Reviewers | Michael Amadi |

> | Status | Review |

Reviewer: once merged this will be the new status
```suggestion
| Status | Implementing Actions |
```

    - Likelihood: Low. This is possible if the deployer engages in poor practices or lacks opsec. Two examples of this are:
        - Custom factories that utilize CREATE and share the same address are used, where tokens are deployed using the same bytecode and nonce.
        - Proxy contracts where the admin is compromised or the proxy address front-runned.
- **Mitigation**: For developers, employ the appropriate deterministic deployment tools, such as the one at `create2Deployer`. In the case of permissioned deployments, common for proxy contracts, ensure to maintain control over those deployments and ownership.
- **Detection**: Support tickets filed by users reporting such issues.
- **Recovery Path(s)**: Pause the token contract where possible and proceed to redeploy and migrate userbase if it is needed.

## Action Items

The following action items need to be done:

- [ ]  Resolve all the comments.
- [ ]  FM1: Implement tests for `SuperchainTokenBridge` predeploy. Currently, the L2 Genesis project is under development and it should take care of these tests.

> - [ ]  FM1: Run off-chain scripts and monitoring measurements (OP Labs).

Reviewer 3: Adding some specifics here about monitoring would help with a definition of done

- [ ]  FM2, FM3, FM4: Communicate to bridge frontends to ensure they prevent users from sending tokens to a chain where the token doesn’t exist.
- [ ]  FM4, FM5: Ensure that the documentation for SuperchainERC20 developers explains the need for deterministic deployment and how to achieve it, including that constructor args affect the resulting address, requiring consistency across chains.

> - [ ]  Ensure the support team is aware of these failure modes and prepared to respond.

Reviewer: We should add some specifics here, we need to figure out who the support team is and what "prepared to respond" really means (e.g. write runbooks)

Author: I have added one extra action item for FM1 and added some details for the case of the team responsible of user support tickets
